File: problems/custom_example/matlab_problems/custom4.m
function problem = custom4()
    % This is a toy example to show how to construct a .m file that returns
    % a instance of Problem class representing a linearly constrained problem.

    p_struct.name = 'custom2';
    p_struct.x0 = [1; 1];
    p_struct.fun = @fun;
    p_struct.xl = [-5; -5];
    p_struct.xu = [5; 5];
    p_struct.aub = [1, 1];
    p_struct.bub = 1;
    p_struct.aeq = [1, -1];
    p_struct.beq = 0;
    p_struct.cub = @cub;
    p_struct.ceq = @ceq;

    problem = Problem(p_struct);
end

function fx = fun(x)
    fx = sum(x.^2);
end

function cubx = cub(x)
    cubx = -sin(x);
end

function ceqx = ceq(x)
    ceqx = sum(x.^4);
end

File: problems/custom/python_problems/custom4.py
import numpy as np

def custom4():
    """
    This is a toy example to show how to construct a .py file that returns a dictionary that describes an nonlinearly constrained optimization problem.
    """
    p_dict = {}
    p_dict['name'] = 'custom4'
    p_dict['x0'] = [1.0, 1.0]
    p_dict['fun'] = fun
    p_dict['xl'] = [-5.0, -5.0]
    p_dict['xu'] = [5.0, 5.0]
    p_dict['aub'] = [1.0, 1.0]
    p_dict['bub'] = 1
    p_dict['aeq'] = [1.0, -1.0]
    p_dict['beq'] = 0
    p_dict['cub'] = cub
    p_dict['ceq'] = ceq

    return p_dict

def fun(x):
    return sum(xi**2 for xi in x)

def cub(x):
    return -np.sin(x)

def ceq(x):
    return sum(xi**4 for xi in x)